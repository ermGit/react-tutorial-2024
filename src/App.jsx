import {UserProfile} from "./components/UserProfile.jsx";
import {UserDetails} from "./components/UserDetails.jsx";
import LoginForm from "./components/LoginForm.jsx";
import {RegisterForm} from "./components/RegisterForm.jsx";
import {useState, useEffect} from "react";

export default function App() {
    const isAuthenticated = true;
    const [username, setUsername] = useState("");
    const [email, setEmail] = useState("");
    const [counter, setCounter] = useState(5);

    const [secondCounter, setSecondCounter] = useState(0);
    const [sync, setSync] = useState(false);

    const [blogPostData, setBlogPostData] = useState(
        {
            title: "",
            body: "",
        }
    );

    // Called when the component is mounted/rendered
    // Can be called many times especially when state changes and causes the
    // component to re-render
    // With [] empty dependency array, the useEffect is only called once.
    // With something in the [counter], then useEffect is called
    // whenever that state is changed.
    useEffect(
        () => {
            console.log("Rendering...");
            document.title = "React Tutorial";
        },
        [
            sync
        ]
    );

    useEffect(() => {
        console.log("Fetching...");
        fetch(
            "https://jsonplaceholder.typicode.com/users",
            {
                method: "GET",
            })
            .then(
                (response) => {
                    console.log(response);
                    return response.json();
                }
            )
            .then(
                (json) => {
                    console.log(json)
                }
            )
            .catch(
                (err) => {
                    console.log(err);
                }
            )
    }, []);

    useEffect(
        () => {
            const controller = new AbortController();
            async function fetchUsers() {
                // The await keyword blocks the script until the call returns
                try {
                    const response = await fetch(
                        "https://jsonplaceholder.typicode.com/users",
                        {signal: controller.signal}
                    );
                    const json = await response.json();
                    console.log("Fetching2...");
                    console.log(json);
                } catch (error) {
                    console.log(error);
                }
            }
            fetchUsers();

            // Called when componenet unmounts
            // Clean up function
            return () => {
                controller.abort();
            };
        }
    );

    useEffect(
        () => {
            const controller = new AbortController();
        }
    );

            const [mockUsers, setUsers] = useState([
        {
            id: 1,
            username: 'John1',
            email: 'john1@example.com',
        },
        {
            id: 2,
            username: 'John2',
            email: 'john2@example.com',
        },
        {
            id: 3,
            username: 'John3',
            email: 'john3@example.com',
        },
        {
            id: 4,
            username: 'John4',
            email: 'john4@example.com',
        }
    ]);

    return isAuthenticated ? (
        <div>
            <form
                onSubmit={
                (e) => {
                    e.preventDefault();
                    if (blogPostData.title && blogPostData.body) {
                        console.log("blogPostData fetching post", blogPostData);
                        fetch(
                            "https://jsonplaceholder.typicode.com/posts",
                            {
                                method: "POST",
                                body: JSON.stringify({
                                    userId: 1,
                                    title: blogPostData.title,
                                    body: blogPostData.body,
                                }),
                                headers: {
                                    "Content-Type": "application/json; charset=UTF-8",
                                },
                            }
                        )
                        .then(response =>
                                response.json()
                        )
                        .then((json) => {
                               console.log("Success!");
                                console.log(json);
                            }
                        )
                        .catch((err) => {
                                console.log(err);
                            }
                        );
                    }
                }}
            >
                <label htmlFor="title">Title</label>
                <br/>
                <input
                    type="text"
                    id="title"
                    value={blogPostData.title}
                    onChange={(e) => {
                        setBlogPostData((currentBlogPostData) =>
                            (
                                {...currentBlogPostData, title: e.target.value}
                            )
                        );
                    }}
                />
                <br/>
                <label htmlFor="body">Body</label>
                <br/>
                <input
                    type="text"
                    id="body"
                    value={blogPostData.body}
                    onChange={(e) => {
                        setBlogPostData((currentBlogPostData) =>
                            (
                                {...currentBlogPostData, body: e.target.value}
                            )
                        );
                    }}
                />
                <br/>
                <button type="submit">
                    Create Post
                </button>
            </form>
            <br/>

            <div>
                <div>You clicked the button {secondCounter} times</div>
                <button
                    onClick={() => {
                        setSecondCounter((secondCounter) => secondCounter + 1)
                    }}
                >
                    Click Me
                </button>
                <button
                    onClick={() => {
                        setSync((current) => !current);
                    }}
                >
                    Sync
                </button>

            </div>
            <br/>

            <form
                onSubmit={(e) => {
                    e.preventDefault();
                    console.log(username);
                    console.log(email);
                    const newUser = {
                        id: counter,
                        username: username,
                        email: email,
                    };
                    setCounter(
                        (currentCounter) => currentCounter + 1
                    );
                    setUsers(
                        (currentUsersState) => [...currentUsersState, newUser]
                    );
                }}
            >
                <div>
                    <label htmlFor="username">Username</label>
                    <br/>
                    <input
                        id="username"
                        name="username"
                        value={username}
                        onChange={(e) => {
                            setUsername(e.target.value);
                        }}
                    />
                </div>
                <div>
                    <label htmlFor="email">Email</label>
                    <br/>
                    <input
                        id="email"
                        name="email"
                        value={email}
                        onChange={(e) => {
                            setEmail(e.target.value);
                        }}
                    />
                </div>
                <button type="submit">Add User</button>
            </form>
            <br/>
            {mockUsers.map(
                (user) => {
                    return (
                        <UserDetails
                            key={user.id}
                            setUsers={setUsers}
                            user={user}
                        />
                    );
                }
            )}

            <h1>Root Component</h1>
            <UserProfile
                username="Bob"
                age={25}
                favoriteFoods={[
                    {
                        name: "Sushi",
                        id: "sushi",
                    },
                    {
                        name: "Pizza",
                        id: "pizza",
                    }
                ]}
            />
        </div>
    ) : (
        <div>
            <h1>
                Please Log in
            </h1>
            <RegisterForm/>
        </div>
    );
}