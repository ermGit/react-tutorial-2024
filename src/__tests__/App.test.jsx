import {describe, it, test, expect} from 'vitest';
import {within} from "@testing-library/dom";
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

const usersData = [
    {
        id: 1,
        username: 'John1',
        email: 'john1@example.com',
    },
    {
        id: 2,
        username: 'John2',
        email: 'john2@example.com',
    },
    {
        id: 3,
        username: 'John3',
        email: 'john3@example.com',
    },
    {
        id: 4,
        username: 'John4',
        email: 'john4@example.com',
    }
];

describe('App', () => {
    it('should render App correctly', async ({expect}) => {
        const { container } = render(<App usersData={[]} />);
        expect(container).toMatchSnapshot();
    });

    describe("When there is only one User", () => {
        describe('Edit Button is Clicked', async () => {
            it('should render Save button', async ({expect}) => {
                const { container } = render(<App usersData={[usersData[0]]} />);
                const editButton = screen.getAllByRole('button', { name: `Edit` });
                await userEvent.click(editButton[0]);
                const saveButton = screen.getByRole('button', { name: `Save` });
                expect(saveButton).toBeInTheDocument();
                //expect(container).toMatchSnapshot();
            });

            it('should display username & email input fields', async ({expect}) => {
                const { container } = render(<App usersData={usersData} />);
                const editButton = screen.getAllByRole('button', { name: `Edit` });

                // Could also be used, but it has lower priority than getbylabel or getbyrole
                /*
                expect(
                    screen.getByTestId("edit-user-button-1")
                ).toBeInTheDocument();
                */

                await userEvent.click(editButton[0]);
                // Added "aria-label="username"" to the input element to get "username" as the name
                expect(
                    screen.getByRole('textbox', { name: `username` })
                ).toBeInTheDocument();

                expect(
                    screen.getByRole('textbox', { name: `email` })
                ).toBeInTheDocument();

                expect(
                    screen.getByLabelText("Username:")
                ).toBeInTheDocument();

                expect(
                    screen.getByLabelText("Email:")
                ).toBeInTheDocument();
            });
        });
    });

    describe("When there are two Users", () => {
        it('should have two users displayed', () => {
            const {container} = render(<App usersData={[usersData[0], usersData[1]]} />);
            const userNames = screen.getAllByText(usersData[0].username);
            expect(userNames).toHaveLength(2);

            expect(screen.getAllByText(usersData[1].username)).toHaveLength(2);
        });

        describe('Edit Button of the first user is Clicked', async () => {
            it('should render Save button', async ({expect}) => {
                const {container} = render(<App usersData={[usersData[0], usersData[1]]} />);

                const userDetails = screen.getAllByTestId("user-details-1");
                expect(userDetails[0]).toBeInTheDocument();

                const editButton = within(userDetails[0]).getByRole('button', { name: `Edit` });
                await userEvent.click(editButton);

                //const editButton = screen.getAllByRole('button', {name: `Edit`});
                //await userEvent.click(editButton[0]);
                const saveButton = within(userDetails[0]).getByRole('button', { name: `Save` });
                expect(saveButton).toBeInTheDocument();
                //expect(container).toMatchSnapshot();
            });

            it('should display username & email input fields', async ({expect}) => {
                const {container} = render(<App usersData={[usersData[0], usersData[1]]}/>);
                const editButton = screen.getAllByRole('button', {name: `Edit`});

                // Could also be used, but it has lower priority than getbylabel or getbyrole
                /*
                expect(
                    screen.getByTestId("edit-user-button-1")
                ).toBeInTheDocument();
                */

                await userEvent.click(editButton[0]);
                // Added "aria-label="username"" to the input element to get "username" as the name
                expect(
                    screen.getByRole('textbox', {name: `username`})
                ).toBeInTheDocument();

                expect(
                    screen.getByRole('textbox', {name: `email`})
                ).toBeInTheDocument();

                expect(
                    screen.getByLabelText("Username:")
                ).toBeInTheDocument();

                expect(
                    screen.getByLabelText("Email:")
                ).toBeInTheDocument();
            });
        });
    });
});

