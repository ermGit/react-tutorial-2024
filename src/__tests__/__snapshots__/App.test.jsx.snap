// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render App correctly 1`] = `
<div>
  <div>
    <div />
    <nav
      class="navbar navbar-expand-lg navbar-dark bg-dark"
    >
      <div
        class="container"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          />
          <li
            class="nav-item"
          />
          <li
            class="nav-item"
          />
        </ul>
      </div>
    </nav>
    <div />
    <br />
    <div>
      <label
        for="data"
      >
        Enter Data
      </label>
      <br />
      <input
        id="data"
        type="text"
      />
    </div>
    <br />
    <div>
      Loading...
    </div>
    <br />
    <form>
      <label
        for="title"
      >
        Title
      </label>
      <br />
      <input
        id="title"
        type="text"
        value=""
      />
      <br />
      <label
        for="body"
      >
        Body
      </label>
      <br />
      <input
        id="body"
        type="text"
        value=""
      />
      <br />
      <button
        type="submit"
      >
        Create Post
      </button>
    </form>
    <br />
    <div>
      <div>
        You clicked the button 
        0
         times
      </div>
      <button>
        Click Me
      </button>
      <button>
        Sync
      </button>
    </div>
    <br />
    <form>
      <div>
        <label
          for="username"
        >
          Username
        </label>
        <br />
        <input
          id="username"
          name="username"
          value=""
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email
        </label>
        <br />
        <input
          id="email"
          name="email"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add User
      </button>
    </form>
    <br />
    <h1>
      Root Component
    </h1>
    <div
      id="user-profile"
    >
      <b
        class="_username_ccaf0b"
        style="font-size: 32px; color: red; font-family: cursive;"
      >
        Username:
      </b>
      <span>
        Bob
      </span>
      <b>
        Age:
      </b>
      <span>
        25
      </span>
      <div>
        <span>
          Email: 
        </span>
        <span>
          bob@bob.com 
        </span>
      </div>
      <section>
        <span
          class="_foodsTitle_38af46"
        >
          Favorite Foods:
        </span>
        <br />
        <ul>
          <li>
            Sushi
          </li>
          <li>
            Pizza
          </li>
          <li>
            Mediterranean Food
          </li>
        </ul>
      </section>
    </div>
  </div>
</div>
`;
